client/driver/docker.go

```diff
@@ -355,35 +360,42 @@ func (d *DockerDriver) FSIsolation() cstructs.FSIsolation {
 	return cstructs.FSIsolationImage
 }
 
-func (d *DockerDriver) Prestart(ctx *ExecContext, task *structs.Task) error {
+func (d *DockerDriver) Prestart(ctx *ExecContext, task *structs.Task) (*CreatedResources, error) {
 	driverConfig, err := NewDockerDriverConfig(task, d.taskEnv)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	// Set state needed by Start()
+	d.driverConfig = driverConfig
+
 	// Initialize docker API clients
 	client, _, err := d.dockerClients()
 	if err != nil {
-		return fmt.Errorf("Failed to connect to docker daemon: %s", err)
+		return nil, fmt.Errorf("Failed to connect to docker daemon: %s", err)
 	}
 
-	if err := d.createImage(driverConfig, client, ctx.TaskDir); err != nil {
-		return err
+	// Ensure the image is available
+	pulled, err := d.createImage(driverConfig, client, ctx.TaskDir)
```

Comment:
I think you are going to want to create a resource regardless of if you pulled the image. If there are two containers using the same image, when the first one exits (the one that pulled it first), the delete image will fail since there is another container using it. So the second one to be started needs to attempt to clean it up.